modules/sdk/src/main/scala/org/tessellation/sdk/http/p2p/middlewares/PeerAuthMiddleware.scala

```diff
@@ -0,0 +1,216 @@
+package org.tessellation.sdk.http.p2p.middleware
+
+import java.security.{PrivateKey, PublicKey}
+
+import cats.data.{Kleisli, OptionT}
+import cats.effect.{Async, Ref, Resource}
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+
+import org.tessellation.schema.cluster.TokenValid
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.cluster.services.Session
+import org.tessellation.sdk.domain.cluster.storage.SessionStorage
+import org.tessellation.sdk.http.p2p.headers.{`X-Id`, `X-Session-Token`}
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signing
+
+import com.comcast.ip4s.{Host, Port}
+import fs2.{Chunk, Stream}
+import org.http4s.Status.Unauthorized
+import org.http4s._
+import org.http4s.client.Client
+import pl.abankowski.httpsigner.http4s._
+import pl.abankowski.httpsigner.signature.generic.{GenericGenerator, GenericVerifier}
+import pl.abankowski.httpsigner.signature.{Generator, Verifier}
+import pl.abankowski.httpsigner.{HttpCryptoConfig, SignatureValid}
+
+object PeerAuthMiddleware {
+
+  private def unauthorized[F[_]] = Response[F](status = Unauthorized)
+
+  private def getOwnTokenHeader[F[_]: Async](sessionStorage: SessionStorage[F]): F[Option[`X-Session-Token`]] =
+    sessionStorage.getToken.map(_.map(t => `X-Session-Token`(t)))
+
+  private def getHostPort[F[_]](req: Request[F]): Option[(Host, Port)] =
```

Comment:
Maybe we could call it `extractHostPort` to differentiate it from functions getting something from storage. Minor thing, just thinking out loud.

Reply:
It gets host and port from the request :)

Reply:
`get` isn't strictly assigned to fetching something from the database

Reply:
I know. But for me still it rather extracts. Though it's not a problem, `get` is also fine.